workspace/notebook/py_casework_local_plan_standardised_post_validation_check.json

```diff
@@ -0,0 +1,652 @@
+					"\n",
+					"def check_ldfno_format(ldfno_value):\n",
+					"    \"\"\"Validates LDFNo field - can only have numbers or 'NA' as Free Text OR NULL values\n",
+					"    Cannot have two numbers concatenated with '&', '/', ',' or ' '\n",
+					"    Returns error description or None if valid\"\"\"\n",
+					"    if not ldfno_value or str(ldfno_value).strip() == \"\":\n",
+					"        return None  # NULL is allowed\n",
+					"    \n",
+					"    ldfno_str = str(ldfno_value).strip()\n",
+					"    \n",
+					"    # Check if it's exactly \"NA\" (case insensitive) - this is allowed\n",
+					"    if ldfno_str.upper() == \"NA\":\n",
+					"        return None\n",
+					"    \n",
+					"    # Check for concatenated numbers with separators\n",
+					"    if re.search(r'\\d+\\s*[&/,\\s]\\s*\\d+', ldfno_str):\n",
+					"        return \"LDFNo contains concatenated numbers with separator (&, /, comma, or space)\"\n",
+					"    \n",
+					"    # Check if it's only numbers (valid)\n",
+					"    if re.match(r'^\\d+$', ldfno_str):\n",
+					"        return None\n",
+					"    \n",
+					"    # If it contains any non-numeric characters (except 'NA' which we already handled)\n",
+					"    return \"LDFNo must contain only numbers or 'NA'\"\n",
+					"\n",
+					"def time_diff_seconds(start, end):\n",
+					"    try:\n",
+					"        if not start or not end:\n",
+					"            return 0\n",
+					"\n",
+					"        # Parse strings into datetime objects if needed\n",
+					"        if isinstance(start, str):\n",
+					"            start = datetime.strptime(start, \"%Y-%m-%d %H:%M:%S.%f\")\n",
+					"        if isinstance(end, str):\n",
+					"            end = datetime.strptime(end, \"%Y-%m-%d %H:%M:%S.%f\")\n",
+					"\n",
+					"        diff_seconds = int((end - start).total_seconds())\n",
+					"        return diff_seconds if diff_seconds > 0 else 0\n",
+					"\n",
+					"    except Exception as e:\n",
+					"        return 0\n",
+					"\n",
+					"#This funtion handles datetime object and covert into string\n",
+					"def datetime_handler(obj):\n",
+					"    if isinstance(obj, datetime):\n",
+					"        return obj.isoformat()\n",
+					"    raise TypeError(f\"Object of type {obj.__class__.__name__} is not JSON serializable\")\n",
+					"    \n",
+					"# Register UDFs\n",
+					"validate_date_udf = udf(check_date_format_columns, StringType())\n",
+					"validate_ldfno_udf = udf(check_ldfno_format, StringType())\n",
+					""
+				],
+				"execution_count": null
+			},
+			{
+				"cell_type": "markdown",
+				"metadata": {
+					"nteract": {
+						"transient": {
+							"deleting": false
+						}
+					}
+				},
+				"source": [
+					"##### Main Execution for data quality checks"
+				]
+			},
+			{
+				"cell_type": "code",
+				"source": [
+					"try:\n",
+					"    # Post Validation Checks processing\n",
+					"\n",
+					"    #Read Source delta table\n",
+					"    src_df = spark.sql(f\"SELECT * FROM {source_delta_table}\")\n",
```

Comment:
You read table into src_df, but the duplicate check uses a separate spark.sql(...) against the table again. Prefer operating on src_df or at least keep one logical plan.